blog/2018/12/22/use-carton-for-your-mojolicious-app-deployment/index.markdown

```diff
@@ -0,0 +1,155 @@
+
+Both 7.77 and 7.90 versions are in our range, but our application does not work on the production server… we need to make the production environment as identical as possible as the development one.
+
+For that, since we have a `cpanfile.snapshot` file from our development server, we can do:
+
+    carton install --deployment
+
+### Features
+
+Per default, `carton install` will install all the *features* dependencies, but we can deactivate some:
+
+    carton install --deployment --without-feature mysql
+
+In order to provide the correct version for all modules, even the optional ones, do a `carton install` on the development server, and use `--without-feature` only while deploying your application: you need to have a `cpanfile.snapshot` containing all modules.
+
+### Start your application
+
+In order to be able to use the `local` directory containing the dependencies, you can prefix your commands with `carton exec`.
+So, to start a Mojolicious application with the built-in server [hypnotoad](https://mojolicious.org/perldoc/Mojo/Server/Hypnotoad), do:
+
+    carton exec hypnotoad script/my_application
+
+That works for all that you can do with your application. Example:
+
+    carton exec script/my_application routes
+
+### Bundling the dependencies
+
+To make installation quicker, carton can bundle all the tarballs for your dependencies into a directory so that you can even install dependencies that are not available on CPAN, such as internal distribution aka DarkPAN:
+
+    carton bundle
+
+This will bundle the tarballs in `vendor/cache`.
+You can now install your dependencies with:
+
+    carton install --cached
+
+Combined with `--deployment` option, you can avoid querying for a database like CPAN Meta DB or downloading files from CPAN mirrors.
+
+You may even avoid the need to install Carton on the production server (but then you will need to add the `local` directory to `@INC` to start your application, since you can’t use `carton exec`):
+
+    cpanm -L local --from "$PWD/vendor/cache" --installdeps --notest --quiet .
+
+## Conclusion
+
+Carton and cpanfile are a great way to ease Mojolicious apps deployment.
+Not only it avoids to list all the dependencies needed by your application in the README or the INSTALL file, but it speeds up deployments and make them more safer, since it sure lowers the risks of bugs due to bad versions of dependencies.
+
+Just be careful to list non-Perl dependencies in README, like `libpq-dev` for [`Mojo::Pg`](https://mojolicious.org/perldoc/Mojo/Pg) 😉
```

Comment:
Can this go up in the cpanfile section? It feels both like (a) it belongs up there more and (b) that as the last sentence in the article, being cautionary, you steal some of your own thunder.

Reply:
:+1: